File: python/hvg.v8.py
from __future__ import annotations
import argparse
import json
import sys
import h5py
import numpy as np
from pathlib import Path
from typing import Any, Optional

## Constants
DEFAULT_NP_DTYPE       = np.float64
LOOM_COMPRESSION       = "gzip"
LOOM_COMPRESSION_LEVEL = 2
OUTPUT_JSON_NAME       = "output.json"

_CELL_ID_PATH = "/col_attrs/CellID"
_GENE_ID_PATH = "/row_attrs/_StableID"

## Error handling
class ErrorJSON(Exception):
    def __init__(self, message: str, output_path: str = None):
        super().__init__(message)
        payload = {"displayed_error": message}
        if output_path:
            with open(output_path, "w", encoding="utf-8") as f:
                json.dump(payload, f, ensure_ascii=False)
        else:
            print(json.dumps(payload, ensure_ascii=False), file=sys.stdout)
        sys.exit(1)

## JSON serializer
def _json_default(obj: Any):
    if isinstance(obj, np.bool_):    return bool(obj)
    if isinstance(obj, np.integer):  return int(obj)
    if isinstance(obj, np.floating): return float(obj)
    if isinstance(obj, np.ndarray):  return obj.tolist()
    return str(obj)

## LOOM helpers
def _read_string_dataset(f: h5py.File, path: str) -> list[str]:
    raw = f[path][()]
    if isinstance(raw, (bytes, np.bytes_)):
        return [raw.decode("utf-8", errors="replace")]
    return [v.decode("utf-8", errors="replace") if isinstance(v, (bytes, np.bytes_)) else str(v) for v in raw.ravel()]

def _write_dataset_1d_int(f_out: h5py.File, path: str, data: np.ndarray) -> int:
    """Write a 1D integer array (0/1 HVG flags). Returns on-disk size in bytes."""
    n = len(data)
    if path in f_out:
        del f_out[path]
    grp = "/".join(path.split("/")[:-1])
    if grp:
        f_out.require_group(grp)
    ds = f_out.create_dataset(path, data=data.astype(np.int32, copy=False),
                               shape=(n,), chunks=(min(n, 1024),),
                               compression=LOOM_COMPRESSION, compression_opts=LOOM_COMPRESSION_LEVEL)
    return int(ds.id.get_storage_size())

def _open_loom_with_retry(path: str, mode: str, max_wait: int = 600) -> tuple:
    """Try to open a LOOM/HDF5 file, retrying every second if locked.
    Returns (h5py.File, wait_time_seconds)."""
    import time as _time
    elapsed = 0
    while True:
        try:
            return h5py.File(path, mode), elapsed
        except OSError as e:
            if elapsed >= max_wait:
                ErrorJSON(f"Could not open {path!r} in mode {mode!r} after {max_wait}s: {e}")
            _time.sleep(1)
            elapsed += 1

## Main HVG function
def hvg(args):
    input_path = Path(args.f).resolve()
    if not input_path.is_file():
        ErrorJSON(f"Input LOOM file not found: {args.f}")

    out_dir = Path(args.o).resolve() if args.o else None
    if out_dir:
        out_dir.mkdir(parents=True, exist_ok=True)
    output_json_path = str(out_dir / OUTPUT_JSON_NAME) if out_dir else None
    plot_dir         = out_dir if out_dir else input_path.parent
    plot_png_path    = str(plot_dir / "plot.hvg.png")
    plot_json_path   = str(plot_dir / "plot.hvg.json")

    if not args.output_meta.startswith("/row_attrs/"):
        ErrorJSON(f"--output_meta must be under /row_attrs/ (e.g. /row_attrs/highly_variable), got: {args.output_meta!r}")

    try:
        import scanpy as sc
        from importlib.metadata import version as _pkg_version
        tool_version = _pkg_version("scanpy")
    except ImportError:
        ErrorJSON("scanpy is not installed. Install it with: pip install scanpy")

    warnings: list[str] = []

    result: dict = {
        "parameters": {},
        "hvg":        {"tool": "scanpy", "tool_version": tool_version,
                       "method": args.method, "n_top_genes": args.n_top_genes, "n_hvg_found": -1},
        "metadata":   [],
        "plots":      [],
        "wait_time":  0,
    }

    f_in, wt = _open_loom_with_retry(str(input_path), "r")
    result["wait_time"] += wt
    with f_in:
        src_path = args.input_meta
        if src_path not in f_in:
            ErrorJSON(f"Source path {src_path!r} not found in LOOM file.")
        node = f_in[src_path]
        if not isinstance(node, h5py.Dataset) or len(node.shape) != 2:
            ErrorJSON(f"Expected a 2-D dataset at {src_path}, got shape {getattr(node, 'shape', '?')}")

        s0, s1 = int(node.shape[0]), int(node.shape[1])
        ra_len = f_in[_GENE_ID_PATH].shape[0]
        ca_len = f_in[_CELL_ID_PATH].shape[0]
        if s0 == ca_len and s1 == ra_len:
            n_genes, n_cells = s1, s0
            matrix_gxc = node[:, :].astype(DEFAULT_NP_DTYPE).T
        else:
            n_genes, n_cells = s0, s1
            matrix_gxc = node[:, :].astype(DEFAULT_NP_DTYPE)

        gene_ids = _read_string_dataset(f_in, _GENE_ID_PATH)
        cell_ids = _read_string_dataset(f_in, _CELL_ID_PATH)

    result["parameters"] = {
        "loom_path":        str(input_path),
        "input_loom_path":  args.input_meta,
        "output_loom_path": args.output_meta,
        "method":           args.method,
        "n_top_genes":      args.n_top_genes,
        "min_mean":         args.min_mean         if args.method in ("seurat", "seurat_v3", "cell_ranger") else None,
        "max_mean":         args.max_mean         if args.method in ("seurat", "seurat_v3", "cell_ranger") else None,
        "min_disp":         args.min_disp         if args.method in ("seurat", "seurat_v3", "cell_ranger") else None,
        "span":             args.span             if args.method == "seurat_v3"                            else None,
        "n_bins":           args.n_bins,
        "batch_key":        args.batch_key,
    }

    # Build AnnData (cells x genes) and run sc.pp.highly_variable_genes
    import anndata
    adata = anndata.AnnData(X=matrix_gxc.T.astype(DEFAULT_NP_DTYPE))
    adata.obs_names = list(cell_ids)
    adata.var_names = list(gene_ids)

    try:
        sc.pp.highly_variable_genes(
            adata,
            n_top_genes = args.n_top_genes,
            flavor      = args.method,
            min_mean    = args.min_mean,
            max_mean    = args.max_mean,
            min_disp    = args.min_disp,
            span        = args.span,
            n_bins      = args.n_bins,
            batch_key   = args.batch_key,
            inplace     = True,
        )
    except (ValueError, ImportError) as e:
        ErrorJSON(f"sc.pp.highly_variable_genes failed with method '{args.method}': {e}. "
                  f"For 'seurat_v3' this usually means skmisc is missing or has a binary incompatibility "
                  f"with the installed numpy version (try: pip install --upgrade skmisc).")

    # scanpy adds adata.var["highly_variable"] as a boolean column
    hvg_flags   = adata.var["highly_variable"].values.astype(np.int32)
    n_hvg_found = int(hvg_flags.sum())
    result["hvg"]["n_hvg_found"] = n_hvg_found

    # Append HVG flags into the LOOM
    f_rw, wt = _open_loom_with_retry(str(input_path), "r+")
    result["wait_time"] += wt
    with f_rw:
        out_path = args.output_meta
        if out_path in f_rw:
            warnings.append(f"Path '{out_path}' already exists and will be overwritten.")
        size_hvg = _write_dataset_1d_int(f_rw, out_path, hvg_flags)

    ## ── Generate plots ───────────────────────────────────────────────────────
    try:
        import plotly.graph_objects as go
        import plotly.io as pio

        var_df = adata.var.copy()
        # Determine y-axis column based on method
        if args.method == "seurat_v3":
            y_col, y_label = "variances_norm",    "Normalized variance"
        else:
            y_col, y_label = "dispersions_norm",  "Normalized dispersion"
        x_label = "Mean expression"

        hvg_mask    = var_df["highly_variable"].values
        x_all       = var_df["means"].values
        y_all       = var_df[y_col].values
        names_all   = var_df.index.tolist()

        fig = go.Figure()
        # Non-HVG trace
        non_mask = ~hvg_mask
        fig.add_trace(go.Scattergl(
            x=x_all[non_mask], y=y_all[non_mask],
            mode="markers", name="Other genes",
            text=[names_all[i] for i in range(len(names_all)) if non_mask[i]],
            hovertemplate="%{text}<br>mean=%{x:.3f}<br>" + y_label + "=%{y:.3f}<extra></extra>",
            marker=dict(color="#CCCCCC", size=3, opacity=0.6)
        ))
        # HVG trace
        fig.add_trace(go.Scattergl(
            x=x_all[hvg_mask], y=y_all[hvg_mask],
            mode="markers", name="HVG",
            text=[names_all[i] for i in range(len(names_all)) if hvg_mask[i]],
            hovertemplate="%{text}<br>mean=%{x:.3f}<br>" + y_label + "=%{y:.3f}<extra></extra>",
            marker=dict(color="#E87C3E", size=5, opacity=0.9)
        ))

        fig.update_layout(
            title=f"Highly Variable Genes — {args.method} ({n_hvg_found}/{n_genes} selected)",
            xaxis_title=x_label, yaxis_title=y_label,
            template="simple_white", width=800, height=600,
            legend=dict(yanchor="top", y=0.99, xanchor="right", x=0.99)
        )

        # Save plotly JSON
        pio.write_json(fig, plot_json_path)
        result["plots"].append({"path": plot_json_path, "type": "plotly_json"})

        # Save PNG (requires kaleido)
        try:
            pio.write_image(fig, plot_png_path, format="png", width=800, height=600, scale=1.5)
            result["plots"].append({"path": plot_png_path, "type": "png"})
        except Exception as e_png:
            warnings.append(f"PNG export failed (kaleido may not be installed): {e_png}")

    except Exception as e_plot:
        warnings.append(f"Plot generation failed: {e_plot}")

    result["metadata"] = [{"name": args.output_meta, "on": "GENE", "type": "INTEGER",
                            "nber_rows": n_genes, "nber_cols": 1,
                            "dataset_size": size_hvg, "imported": 0}]

    if warnings:
        result["warnings"] = warnings

    json_str = json.dumps(result, default=_json_default, ensure_ascii=False, separators=(",", ":"))
    if output_json_path:
        with open(output_json_path, "w", encoding="utf-8") as fj:
            fj.write(json_str)
    else:
        print(json_str)

## CLI
HELP_TEXT = """
HVG Script — scanpy backend

Options:
  -f             Input LOOM file.                                            [required]
  --input_meta   Path to the count/normalized matrix inside the LOOM.       [required]
  --output_meta  Path in /row_attrs/ where the 0/1 HVG flag is stored.      [required]
                   Example: /row_attrs/highly_variable
  --method       HVG method: seurat | seurat_v3 | cell_ranger               [required]
                   seurat, cell_ranger : expects log-normalized data as --input_meta
                   seurat_v3          : expects raw counts as --input_meta
  -o             Output folder for output.json. [optional, default: stdout]

  -- sc.pp.highly_variable_genes parameters --------------------------------------
  --n_top_genes   Number of top variable genes.              [default: 2000]
  --min_mean      Min mean expression threshold.             [default: 0.0125]
  --max_mean      Max mean expression threshold.             [default: 3.0]
  --min_disp      Min dispersion threshold.                  [default: 0.5]
  --span          Span for loess (seurat_v3 only).           [default: 0.3]
  --n_bins        Number of bins for mean expression.        [default: 20]
  --batch_key     Column in obs for batch correction.        [default: None]

  --help         Show this message and exit.
"""

def main():
    if "--help" in sys.argv:
        print(HELP_TEXT)
        sys.exit(0)

    parser = argparse.ArgumentParser(description="HVG Script (scanpy)", add_help=False)
    parser.add_argument("-f",            required=True)
    parser.add_argument("-o",            required=False, default=None)
    parser.add_argument("--input_meta",  required=True)
    parser.add_argument("--output_meta", required=True)
    parser.add_argument("--method",      required=True, choices=["seurat", "seurat_v3", "cell_ranger"])
    parser.add_argument("--n_top_genes", type=int,   default=2000)
    parser.add_argument("--min_mean",    type=float, default=0.0125)
    parser.add_argument("--max_mean",    type=float, default=3.0)
    parser.add_argument("--min_disp",    type=float, default=0.5)
    parser.add_argument("--span",        type=float, default=0.3)
    parser.add_argument("--n_bins",      type=int,   default=20)
    parser.add_argument("--batch_key",   type=str,   default=None)

    args = parser.parse_args()
    hvg(args)

if __name__ == "__main__":
    main()

File: R/hvg.v8.R
#!/usr/bin/env Rscript
# hvg.R — Highly Variable Gene selection on scRNA-seq data from a LOOM file using Seurat v5.
#
# Reads a LOOM file, identifies HVGs, writes a 0/1 flag per gene into /row_attrs/,
# and writes output.json.
#
# Required packages: hdf5r, Seurat (>= 5.0), jsonlite, Matrix, optparse

## ── Phase 0: HELP (no library needed, runs instantly) ────────────────────────

HELP_TEXT <- "
HVG Script — Seurat v5 backend

Reads a LOOM file, runs Seurat v5 FindVariableFeatures, writes a 0/1 flag
per gene into /row_attrs/ of the input LOOM file, and writes output.json.

Options:
  -f / --file        Input LOOM file.                                       [required]
  --input_meta       Path to the count/normalized matrix inside the LOOM.   [required]
                       Examples: /matrix   /layers/normalized_ln
  --output_meta      Path in /row_attrs/ where the 0/1 HVG flag is stored.  [required]
                       Example: /row_attrs/highly_variable
  --method           HVG method: vst | mvp | disp                           [required]
  -o / --output_dir  Output folder for output.json. [optional, default: stdout]
  --assay            Seurat assay name.              [default: RNA]

  -- FindVariableFeatures parameters (defaults match Seurat v5) -------------------
  --n_top_genes      Number of top variable genes to select.     [default: 2000]
  --loess_span       Span for loess smoothing (vst only).        [default: 0.3]
  --clip_max         After standardization, values above this
                       are clipped (vst only).                   [default: Inf]
  --mean_cutoff      Min and max mean expression cutoffs for
                       mvp/disp, as 'min,max' (e.g. '0.1,8').   [default: 0.1,8]
  --dispersion_cutoff  Min and max dispersion cutoffs for
                       mvp/disp, as 'min,max' (e.g. '1,Inf').   [default: 1,Inf]
  --num_bin          Number of bins for mvp.                     [default: 20]

  --help             Show this message and exit.

Output JSON hvg block:
  tool          str  — always 'Seurat'
  tool_version  str  — installed Seurat version
  method        str  — HVG method used
  n_top_genes   int  — number of genes requested
  n_hvg_found   int  — number of HVGs actually selected

Output JSON metadata entry:
  name          str  — LOOM-internal path of the HVG flag (= --output_meta)
  on            str  — always 'GENE'
  type          str  — always 'INTEGER'
  nber_rows     int  — number of genes
  nber_cols     int  — always 1
  dataset_size  int  — on-disk compressed size in bytes
  imported      int  — always 0

Mandatory parameters: -f/--file, --input_meta, --output_meta, --method
"

argv <- commandArgs(trailingOnly = TRUE)
if (length(argv) == 0 || "--help" %in% argv || "-h" %in% argv) { cat(HELP_TEXT); quit(save = "no", status = 0) }

## ── Phase 1: minimal imports ──────────────────────────────────────────────────

suppressPackageStartupMessages({
  library(optparse)
  library(jsonlite)
})

ErrorJSON <- function(message, output_path = NULL) {
  payload  <- list(displayed_error = message)
  json_str <- toJSON(payload, auto_unbox = TRUE)
  if (!is.null(output_path)) writeLines(json_str, con = output_path) else cat(json_str, "\n")
  quit(save = "no", status = 1)
}

## Constants
LOOM_GZIP_LEVEL  <- 2L
OUTPUT_JSON_NAME <- "output.json"

CELL_ID_PATH <- "col_attrs/CellID"
GENE_ID_PATH <- "row_attrs/_StableID"

## CLI
option_list <- list(
  make_option(c("-f", "--file"),       type = "character", default = NULL,    help = "Input LOOM file. [required]"),
  make_option(c("-o", "--output_dir"), type = "character", default = NULL,    help = "Output folder. [optional, default: stdout]"),
  make_option("--input_meta",          type = "character", default = NULL,    help = "Path to the count/normalized matrix. [required]"),
  make_option("--output_meta",         type = "character", default = NULL,    help = "Path in /row_attrs/ for the 0/1 HVG flag (e.g. /row_attrs/highly_variable). [required]"),
  make_option("--method",              type = "character", default = NULL,    help = "HVG method: vst | mvp | disp. [required]"),
  make_option("--assay",               type = "character", default = "RNA",   help = "Seurat assay name. [default: RNA]"),
  make_option("--n_top_genes",         type = "integer",   default = 2000L,   help = "[FindVariableFeatures] Top variable genes. [default: 2000]"),
  make_option("--loess_span",          type = "double",    default = 0.3,     help = "[FindVariableFeatures/vst] Loess span. [default: 0.3]"),
  make_option("--clip_max",            type = "double",    default = Inf,     help = "[FindVariableFeatures/vst] Clip max. [default: Inf]"),
  make_option("--mean_cutoff",         type = "character", default = "0.1,8", help = "[FindVariableFeatures/mvp|disp] Mean cutoff 'min,max'. [default: 0.1,8]"),
  make_option("--dispersion_cutoff",   type = "character", default = "1,Inf", help = "[FindVariableFeatures/mvp|disp] Dispersion cutoff 'min,max'. [default: 1,Inf]"),
  make_option("--num_bin",             type = "integer",   default = 20L,     help = "[FindVariableFeatures/mvp] Number of bins. [default: 20]")
)

parser <- OptionParser(option_list = option_list, add_help_option = FALSE)
args   <- parse_args(parser)

if (is.null(args$file))        ErrorJSON("Missing required argument -f.")
if (is.null(args$input_meta))  ErrorJSON("Missing required argument --input_meta.")
if (is.null(args$output_meta)) ErrorJSON("Missing required argument --output_meta.")
if (is.null(args$method))      ErrorJSON("Missing required argument --method.")

valid_methods <- c("vst", "mvp", "disp")
if (!args$method %in% valid_methods) ErrorJSON(paste0("Unknown method '", args$method, "'. Valid: ", paste(valid_methods, collapse = ", "), "."))
if (!startsWith(args$output_meta, "/row_attrs/")) ErrorJSON(paste0("--output_meta must be under /row_attrs/ (e.g. /row_attrs/highly_variable), got: '", args$output_meta, "'."))

input_path <- normalizePath(args$file, mustWork = FALSE)
if (!file.exists(input_path)) ErrorJSON(paste0("Input LOOM file not found: ", args$file))

# Parse cutoff pairs
parse_cutoff <- function(s, name) {
  parts <- trimws(strsplit(s, ",")[[1]])
  if (length(parts) != 2) ErrorJSON(paste0("--", name, " must be 'min,max', got: '", s, "'."))
  as.numeric(parts)
}
mean_cutoff        <- parse_cutoff(args$mean_cutoff,       "mean_cutoff")
dispersion_cutoff  <- parse_cutoff(args$dispersion_cutoff, "dispersion_cutoff")

out_dir          <- if (!is.null(args$output_dir)) normalizePath(args$output_dir, mustWork = FALSE) else NULL
if (!is.null(out_dir)) dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
output_json_path <- if (!is.null(out_dir)) file.path(out_dir, OUTPUT_JSON_NAME) else NULL
plot_dir         <- if (!is.null(out_dir)) out_dir else dirname(input_path)
plot_png_path    <- file.path(plot_dir, "plot.hvg.png")
plot_json_path   <- file.path(plot_dir, "plot.hvg.json")

## ── Phase 2: heavy imports ────────────────────────────────────────────────────

suppressPackageStartupMessages({
  library(hdf5r)
  library(Matrix)
  library(Seurat)
  library(ggplot2)
  library(plotly)
})

seurat_version <- tryCatch(as.character(packageVersion("Seurat")), error = function(e) "0.0.0")
if (numeric_version(seurat_version) < numeric_version("5.0.0")) ErrorJSON(paste0("Seurat v5 required (found v", seurat_version, ")."))

## ── LOOM helpers ──────────────────────────────────────────────────────────────

write_1d_integer <- function(h5, path, vec) {
  # Write a 1D integer vector (0/1 HVG flags).
  parts <- Filter(nzchar, strsplit(path, "/")[[1]])
  if (length(parts) > 1) {
    grp <- paste(parts[-length(parts)], collapse = "/")
    if (!h5$exists(grp)) h5$create_group(grp)
  }
  if (h5$exists(path)) {
    pg <- if (length(parts) > 1) h5[[paste(parts[-length(parts)], collapse = "/")]] else h5
    pg$link_delete(parts[length(parts)])
  }
  n  <- length(vec)
  ds <- h5$create_dataset(name = path, dtype = h5types$H5T_NATIVE_INT32, dims = n,
                           chunk_dims = min(n, 1024L), gzip_level = LOOM_GZIP_LEVEL)
  ds[1:n] <- as.integer(vec)
  invisible(ds$get_storage_size())
}

## ── Shared warnings + retry helper ───────────────────────────────────────────

warn_env <- new.env(parent = emptyenv())
warn_env$w <- character(0)

open_loom_with_retry <- function(path, mode, max_wait = 600L) {
  elapsed <- 0L
  repeat {
    h5 <- tryCatch(H5File$new(path, mode = mode), error = function(e) e)
    if (!inherits(h5, "error")) return(list(h5 = h5, wait_time = elapsed))
    if (elapsed >= max_wait) ErrorJSON(paste0("Could not open '", path, "' after ", max_wait, "s: ", conditionMessage(h5)))
    Sys.sleep(1)
    elapsed <- elapsed + 1L
  }
}

capture_warnings <- function(expr) {
  withCallingHandlers(expr, warning = function(w) {
    warn_env$w <- c(warn_env$w, conditionMessage(w))
    invokeRestart("muffleWarning")
  })
}

## ── Result skeleton ───────────────────────────────────────────────────────────

result <- list(
  parameters = list(),
  hvg        = list(tool = "Seurat", tool_version = seurat_version,
                    method = args$method, n_top_genes = args$n_top_genes, n_hvg_found = -1L),
  metadata   = list(),
  plots      = list(),
  wait_time  = 0L
)

## ── Read the LOOM file ────────────────────────────────────────────────────────

ret   <- open_loom_with_retry(input_path, mode = "r")
h5_in <- ret$h5
result$wait_time <- result$wait_time + ret$wait_time

src_path <- sub("^/", "", args$input_meta)
if (!h5_in$exists(src_path)) ErrorJSON(paste0("Source path '", args$input_meta, "' not found in LOOM file."))

raw_mat <- h5_in[[src_path]][,]
ra_len  <- h5_in[[GENE_ID_PATH]]$dims
ca_len  <- h5_in[[CELL_ID_PATH]]$dims
if (!is.na(ra_len) && !is.na(ca_len) && nrow(raw_mat) == ca_len && ncol(raw_mat) == ra_len) {
  matrix_gxc <- t(raw_mat)
} else {
  matrix_gxc <- raw_mat
}
n_genes <- nrow(matrix_gxc)
n_cells <- ncol(matrix_gxc)

cell_ids <- as.character(h5_in[[CELL_ID_PATH]][])
gene_ids <- as.character(h5_in[[GENE_ID_PATH]][])
rownames(matrix_gxc) <- gene_ids
colnames(matrix_gxc) <- cell_ids
h5_in$close_all()

# Represent Inf as the string "Inf" so JSON doesn't serialize it as null
inf_to_str <- function(v) ifelse(is.infinite(v) & v > 0, "Inf", ifelse(is.infinite(v) & v < 0, "-Inf", v))

result$parameters <- list(
  loom_path        = input_path,
  input_loom_path  = args$input_meta,
  output_loom_path = args$output_meta,
  method           = args$method,
  n_top_genes      = args$n_top_genes,
  loess_span       = if (args$method == "vst")             args$loess_span              else NULL,
  clip_max         = if (args$method == "vst")             inf_to_str(args$clip_max)    else NULL,
  mean_cutoff      = if (args$method %in% c("mvp","disp")) mean_cutoff                  else NULL,
  dispersion_cutoff= if (args$method %in% c("mvp","disp")) inf_to_str(dispersion_cutoff) else NULL,
  num_bin          = if (args$method == "mvp")             args$num_bin                 else NULL
)

## ── Create Seurat object and find variable features ───────────────────────────

# vst reads from 'counts'; mvp/disp read from 'data'.
# For vst: load the real matrix as counts (vst needs raw-ish counts).
# For mvp/disp: use a zero sparse fake for counts (faster) and load the real matrix into data.
if (args$method == "vst") {
  sparse_counts <- suppressWarnings(Matrix(matrix_gxc, sparse = TRUE))
  seurat_obj    <- suppressWarnings(CreateSeuratObject(counts = sparse_counts, assay = args$assay, project = "hvg_r"))
} else {
  fake_counts <- Matrix::Matrix(0, nrow = n_genes, ncol = n_cells, sparse = TRUE, dimnames = list(gene_ids, cell_ids))
  seurat_obj  <- suppressWarnings(CreateSeuratObject(counts = fake_counts, assay = args$assay, project = "hvg_r"))
  seurat_obj[[args$assay]] <- SetAssayData(seurat_obj[[args$assay]], layer = "data", new.data = matrix_gxc)
}

seurat_obj <- capture_warnings(suppressMessages(FindVariableFeatures(
  seurat_obj,
  assay              = args$assay,
  selection.method   = args$method,
  nfeatures          = args$n_top_genes,
  loess.span         = args$loess_span,
  clip.max           = args$clip_max,
  mean.cutoff        = mean_cutoff,
  dispersion.cutoff  = dispersion_cutoff,
  num.bin            = args$num_bin,
  verbose            = FALSE
)))

hvg_names   <- VariableFeatures(seurat_obj, assay = args$assay)
hvg_flags   <- as.integer(gene_ids %in% hvg_names)
n_hvg_found <- sum(hvg_flags)

result$hvg$n_hvg_found <- as.integer(n_hvg_found)

## ── Write HVG flags into the LOOM ────────────────────────────────────────────

ret     <- open_loom_with_retry(input_path, mode = "a")
h5_loom <- ret$h5
result$wait_time <- result$wait_time + ret$wait_time

out_path_h5 <- sub("^/", "", args$output_meta)
if (h5_loom$exists(out_path_h5)) {
  warn_env$w <- c(warn_env$w, paste0("Path '", args$output_meta, "' already exists and will be overwritten."))
}
size_hvg <- write_1d_integer(h5_loom, out_path_h5, hvg_flags)
h5_loom$close_all()

## ── Metadata + JSON ──────────────────────────────────────────────────────────

result$metadata <- list(list(
  name         = args$output_meta,
  on           = "GENE",
  type         = "INTEGER",
  nber_rows    = as.integer(n_genes),
  nber_cols    = 1L,
  dataset_size = as.integer(size_hvg),
  imported     = 0L
))

if (length(warn_env$w) > 0) result$warnings <- as.list(warn_env$w)

## ── Generate plots ───────────────────────────────────────────────────────────

tryCatch({
  # Extract per-gene statistics from the Seurat object
  hvf      <- HVFInfo(seurat_obj, assay = args$assay, method = args$method)
  hvf$gene <- rownames(hvf)
  hvf$hvg  <- hvf$gene %in% hvg_names

  # Determine axes based on method — detect column names dynamically since
  # Seurat v5 StdAssay may prefix them (e.g. vst.variance.standardized, mvp.dispersion.scaled)
  hvf$x_val <- hvf[[names(hvf)[grepl("^mean$|[.]mean$", names(hvf))][1]]]
  x_label   <- "Mean expression"
  if (args$method == "vst") {
    y_col   <- names(hvf)[grepl("standardized", names(hvf))][1]
    y_label <- "Standardized variance"
  } else {
    y_col   <- names(hvf)[grepl("normalized|scaled", names(hvf))][1]
    y_label <- "Normalized dispersion"
  }
  if (is.na(y_col)) stop(paste0("Cannot find y-axis column in HVFInfo. Available: ", paste(names(hvf), collapse = ", ")))
  hvf$y_val <- hvf[[y_col]]

  # Build ggplot
  p <- ggplot(hvf, aes(x = x_val, y = y_val, text = gene)) +
    geom_point(data = hvf[!hvf$hvg, ], color = "#CCCCCC", size = 0.4, alpha = 0.6) +
    geom_point(data = hvf[ hvf$hvg, ], color = "#E87C3E", size = 0.8, alpha = 0.9) +
    labs(title   = paste0("Highly Variable Genes — ", args$method,
                          " (", n_hvg_found, "/", n_genes, " selected)"),
         x = x_label, y = y_label) +
    theme_bw(base_size = 11) +
    theme(panel.grid.minor = element_blank())

  # Save PNG
  ggsave(plot_png_path, plot = p, width = 8, height = 6, dpi = 150)

  # Save plotly JSON — plotly:::to_JSON is plotly's own serializer and correctly
  # handles all internal plotly classes (plotly_eval etc.) that jsonlite cannot.
  pl       <- ggplotly(p, tooltip = "text")
  fig_json <- plotly:::to_JSON(plotly::plotly_build(pl)$x)
  writeLines(fig_json, plot_json_path)

  result$plots <- list(
    list(path = plot_png_path,  type = "png"),
    list(path = plot_json_path, type = "plotly_json")
  )
}, error = function(e) {
  warn_env$w <- c(warn_env$w, paste0("Plot generation failed: ", conditionMessage(e)))
})

## ── Emit output.json ─────────────────────────────────────────────────────────

if (length(warn_env$w) > 0) result$warnings <- as.list(warn_env$w)

json_str <- toJSON(result, auto_unbox = TRUE, null = "null", na = "null", pretty = FALSE, digits = 7)
if (!is.null(output_json_path)) writeLines(json_str, con = output_json_path) else cat(json_str, "\n")